cloudmetrics/objects/metrics/iorg.py

```diff
@@ -0,0 +1,296 @@
+        )
+        object_radii = object_diameters / 2.0
+        object_radii = np.flip(np.sort(object_radii))  # Largest to smallest
+        nnd_cdf_rand = _compute_inhibition_nearest_neighbour_distribution(
+            object_radii=object_radii,
+            nn_window=nn_window,
+            domain_shape=domain_shape,
+            dist_bins=dist_bins,
+            **reference_dist_kwargs,
+        )
+    else:
+        raise NotImplementedError(reference_dist)
+
+    iorg_value = np.trapz(nnd_cdf_scene, nnd_cdf_rand)
+
+    return iorg_value
+
+
+def _compute_inhibition_nearest_neighbour_distribution(
+    object_radii,
+    nn_window,
+    domain_shape,
+    dist_bins,
+    max_iterations=100,
+    debug=False,
+    random_seed=None,
+):
+    """
+    Compute a reference nearest-neigbour distance distribution using the the
+    inhibition nearest neighbour method as proposed by Benner & Curry (1998)
+    and detailed in Antonissen (2019). The method attempts to randomly place
+    circular objects (using the provided object radii) in the provided domain
+
+    Parameters
+    ----------
+    object_radii:   numpy array of object radii to attempt to place in the domain
+    max_iterations:  maximum number of iterations to use within the algorithm
+                     before giving up
+    num_placements:  number of times the entire random circle placement routine
+                     is carried out, and thus number of different random
+                     nearest neighbour distance cumulative density functions
+                     are computed. The function output is an average over the
+                     num_placements different values of iorg this generates.
+    dist_bins:       distance bins to use when computing comulative distribution
+                     function for nearest-neighbour distances of placed object
+    """
+    rng = np.random.default_rng(random_seed)
+    # Attempt to randomly place all circles in scene without ovelapping
+    i = 0
+    placed_circles = []
+    place_count = 0
+    while i < len(object_radii) and place_count < max_iterations:
+        new = CloudCircle(object_radii[i], domain_shape, rng)
+        placeable = True
+
+        # If the circles overlap -> Place again
+        if _check_circle_overlap(new, placed_circles):
+            placeable = False
+            place_count += 1
+
+        if placeable:
+            placed_circles.append(new)
+            i += 1
+            place_count = 0
+
+    if place_count == max_iterations:
+        raise Exception("Unable to place circles in this image")
+
+    # Gather positions in array
+    pos_rand = np.zeros((len(placed_circles), 2))
+    for i, placed_circle in enumerate(placed_circles):
+        pos_rand[i, 0] = placed_circle.x
+        pos_rand[i, 1] = placed_circle.y
+
+    # If field has open bcs, do not compute nn distances using
+    # periodic bcs
+    nnd_rand = find_nearest_neighbors(pos_rand, nn_window)
```

Comment:
Actually @martinjanssens, I do have a question here: Do we need to restrict the the maximum distance for the nearest-neighbour search periodic domains? The wrapping you wrote that places all centroids inside the domain should ensure that it is impossible to have any nearest-neighbour pair further apart than the domain size.

Another question I realised while typing this: If I have a set of centroids (two objects) placed like this

```
[      ]
................
................
......x..x......
................
................
................
................
................
................
................
................
................
................
................
................
................
```

After wrapping so the centroids are inside the domain that will result in:

```
[      ]
........
........
.x....x.
........
........
........
........
........
```

Meaning that the wrapping causes these points to move further apart (!) Have a got something wrong here or could this happen?

Reply:
Really great questions, I remember spending quite some time thinking about these things! The way I understand the `KDTree` method operating inside `find_nearest_neighbours` to work, it should handle all this when you give it the `boxsize` argument, i.e. it realises that it should always use the smallest distance between the points to compute the distance to the nearest neighbour and uses this when it's specified (see https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.html). So I think that together with all the points being inside the domain and the domainsize being restricted to half the size of the input mask, everything should work consistently. I remember doing some tests that showed as much too, but admit I can't show anything right away. I'm curious what you think!

Reply:
Thanks for this @martinjanssens! I've created a test to check that the nearest-neighbor distance calculation indeed works for periodic domains (https://github.com/cloudsci/cloudmetrics/pull/61). If you could give that a quick glance that would be great. I'd like to get that merged in, and then I'll merge this pull-request in too, and we'll be ready for our first release!